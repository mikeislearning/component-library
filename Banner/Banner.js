import React from 'react';
import { SafeAreaView, View } from 'react-native';
import { PropTypes } from 'prop-types';
import LinearGradient from 'react-native-linear-gradient';
import { colors } from 'app/theme';

import styles from './Banner.styles';

export default class Banner extends React.Component {
  static propTypes = {
    children: PropTypes.node,
    isOtherUserProfile: PropTypes.bool,
    role: PropTypes.string,
  };

  static defaultProps = {
    isOtherUserProfile: false,
    role: 'user',
  };

  render() {
    const { children, isOtherUserProfile, role } = this.props;
    const isAmbassador = role === 'ambassador';

    const bannerStyles = [styles.banner];
    const containerStyles = [styles.container];
    const circleSmallStyles = [styles.circle, styles.circleSmall];
    const circleMediumStyles = [styles.circle, styles.circleMedium];
    const circleLargeStyles = [styles.circle, styles.circleLarge];

    if (isAmbassador) {
      bannerStyles.push(styles.goldBanner);
      containerStyles.push(styles.goldBanner);
    } else {
      if (isOtherUserProfile) {
        bannerStyles.push(styles.otherBanner);
        containerStyles.push(styles.otherBanner);
        circleSmallStyles.push(styles.circleRed);
        circleMediumStyles.push(styles.circleGray);
        circleLargeStyles.push(styles.circleRed);
      } else {
        bannerStyles.push(styles.redBanner);
        containerStyles.push(styles.redBanner);
      }
    }

    return (
      <SafeAreaView style={containerStyles}>
        <View style={bannerStyles}>
          {isAmbassador && (
            <LinearGradient
              colors={[colors.gold.medium, colors.gold.light]}
              end={{ x: 1, y: 0 }}
              start={{ x: 0, y: 0 }}
              style={styles.linearGradient}
            />
          )}
          <View style={circleSmallStyles} />
          <View style={circleMediumStyles} />
          <View style={circleLargeStyles} />

          <View style={styles.content}>{children}</View>
        </View>
      </SafeAreaView>
    );
  }
}
