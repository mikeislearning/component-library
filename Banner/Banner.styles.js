import { StyleSheet } from 'react-native';
import { colors } from 'app/theme';

const CIRCLE_SMALL = 156;
const CIRCLE_MEDIUM = 226;
const CIRCLE_LARGE = 376;

const styles = StyleSheet.create({
  banner: {
    alignItems: 'flex-end',
    flexDirection: 'row',
    height: 57,
    position: 'relative',
  },
  circle: {
    borderColor: colors.white,
    borderRadius: CIRCLE_LARGE,
    borderWidth: 1,
    left: '50%',
    position: 'absolute',
  },
  circleGray: {
    borderColor: colors.gray.light,
  },
  circleLarge: {
    bottom: -CIRCLE_LARGE / 2,
    height: CIRCLE_LARGE,
    marginLeft: -CIRCLE_LARGE / 2,
    width: CIRCLE_LARGE,
  },
  circleMedium: {
    bottom: -CIRCLE_MEDIUM / 2,
    height: CIRCLE_MEDIUM,
    marginLeft: -CIRCLE_MEDIUM / 2,
    width: CIRCLE_MEDIUM,
  },
  circleRed: {
    borderColor: colors.red.medium,
  },
  circleSmall: {
    bottom: -CIRCLE_SMALL / 2,
    height: CIRCLE_SMALL,
    marginLeft: -CIRCLE_SMALL / 2,
    width: CIRCLE_SMALL,
  },
  container: {
    overflow: 'hidden',
  },
  content: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-between',
    paddingHorizontal: 8,
    zIndex: 1,
  },
  goldBanner: {
    backgroundColor: colors.gold.medium,
  },
  linearGradient: {
    flex: 1,
    height: 100,
    left: 0,
    position: 'absolute',
    top: -43, // This is to cover iPhoneX safe area
    width: '100%',
  },
  otherBanner: {
    backgroundColor: colors.white,
    borderBottomColor: colors.gray.light,
    borderBottomWidth: 1,
  },
  redBanner: {
    backgroundColor: colors.red.medium,
  },
});

export default styles;
