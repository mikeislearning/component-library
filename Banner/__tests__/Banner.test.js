import React from 'react';
import { Text } from 'react-native';
import { render } from 'react-native-testing-library';

import Banner from '../Banner';

describe('<Banner />', () => {
  it('renders correctly', () => {
    const banner = render(
      <Banner isOtherUserProfile={true}>
        <Text>Banner title</Text>
      </Banner>
    );
    expect(banner).toMatchSnapshot();
  });

  it('renders a gold banner for ambassadors', () => {
    const banner = render(
      <Banner role="ambassador">
        <Text>Ambassador Content</Text>
      </Banner>
    );
    expect(banner).toMatchSnapshot();
  });

  it('renders a red banner for own profile pages of regular users', () => {
    const banner = render(
      <Banner isOtherUserProfile={false} role="user">
        <Text>Ambassador Content</Text>
      </Banner>
    );
    expect(banner).toMatchSnapshot();
  });
});
