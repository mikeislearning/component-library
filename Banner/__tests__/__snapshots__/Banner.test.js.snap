// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Banner /> renders a gold banner for ambassadors 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Array [
      Object {
        "overflow": "hidden",
      },
      Object {
        "backgroundColor": "#FFAE00",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "flex-end",
          "flexDirection": "row",
          "height": 57,
          "position": "relative",
        },
        Object {
          "backgroundColor": "#FFAE00",
        },
      ]
    }
  >
    <BVLinearGradient
      colors={
        Array [
          4294946304,
          4294899088,
        ]
      }
      endPoint={
        Object {
          "x": 1,
          "y": 0,
        }
      }
      locations={null}
      startPoint={
        Object {
          "x": 0,
          "y": 0,
        }
      }
      style={
        Object {
          "flex": 1,
          "height": 100,
          "left": 0,
          "position": "absolute",
          "top": -43,
          "width": "100%",
        }
      }
    />
    <View
      style={
        Array [
          Object {
            "borderColor": "#FFFFFF",
            "borderRadius": 200,
            "borderWidth": 1,
            "left": "50%",
            "position": "absolute",
          },
          Object {
            "bottom": -78,
            "height": 156,
            "marginLeft": -78,
            "width": 156,
          },
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "borderColor": "#FFFFFF",
            "borderRadius": 200,
            "borderWidth": 1,
            "left": "50%",
            "position": "absolute",
          },
          Object {
            "bottom": -113,
            "height": 226,
            "marginLeft": -113,
            "width": 226,
          },
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "borderColor": "#FFFFFF",
            "borderRadius": 200,
            "borderWidth": 1,
            "left": "50%",
            "position": "absolute",
          },
          Object {
            "bottom": -188,
            "height": 376,
            "marginLeft": -188,
            "width": 376,
          },
        ]
      }
    />
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "paddingHorizontal": 8,
          "zIndex": 1,
        }
      }
    >
      <Text>
        Ambassador Content
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;

exports[`<Banner /> renders a red banner for own profile pages of regular users 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Array [
      Object {
        "overflow": "hidden",
      },
      Object {
        "backgroundColor": "#CC2E41",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "flex-end",
          "flexDirection": "row",
          "height": 57,
          "position": "relative",
        },
        Object {
          "backgroundColor": "#CC2E41",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "borderColor": "#FFFFFF",
            "borderRadius": 200,
            "borderWidth": 1,
            "left": "50%",
            "position": "absolute",
          },
          Object {
            "bottom": -78,
            "height": 156,
            "marginLeft": -78,
            "width": 156,
          },
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "borderColor": "#FFFFFF",
            "borderRadius": 200,
            "borderWidth": 1,
            "left": "50%",
            "position": "absolute",
          },
          Object {
            "bottom": -113,
            "height": 226,
            "marginLeft": -113,
            "width": 226,
          },
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "borderColor": "#FFFFFF",
            "borderRadius": 200,
            "borderWidth": 1,
            "left": "50%",
            "position": "absolute",
          },
          Object {
            "bottom": -188,
            "height": 376,
            "marginLeft": -188,
            "width": 376,
          },
        ]
      }
    />
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "paddingHorizontal": 8,
          "zIndex": 1,
        }
      }
    >
      <Text>
        Ambassador Content
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;

exports[`<Banner /> renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Array [
      Object {
        "overflow": "hidden",
      },
      Object {
        "backgroundColor": "#FFFFFF",
        "borderBottomColor": "#ECECEC",
        "borderBottomWidth": 1,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "flex-end",
          "flexDirection": "row",
          "height": 57,
          "position": "relative",
        },
        Object {
          "backgroundColor": "#FFFFFF",
          "borderBottomColor": "#ECECEC",
          "borderBottomWidth": 1,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "borderColor": "#FFFFFF",
            "borderRadius": 200,
            "borderWidth": 1,
            "left": "50%",
            "position": "absolute",
          },
          Object {
            "bottom": -78,
            "height": 156,
            "marginLeft": -78,
            "width": 156,
          },
          Object {
            "borderColor": "#CC2E41",
          },
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "borderColor": "#FFFFFF",
            "borderRadius": 200,
            "borderWidth": 1,
            "left": "50%",
            "position": "absolute",
          },
          Object {
            "bottom": -113,
            "height": 226,
            "marginLeft": -113,
            "width": 226,
          },
          Object {
            "borderColor": "#ECECEC",
          },
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "borderColor": "#FFFFFF",
            "borderRadius": 200,
            "borderWidth": 1,
            "left": "50%",
            "position": "absolute",
          },
          Object {
            "bottom": -188,
            "height": 376,
            "marginLeft": -188,
            "width": 376,
          },
          Object {
            "borderColor": "#CC2E41",
          },
        ]
      }
    />
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "paddingHorizontal": 8,
          "zIndex": 1,
        }
      }
    >
      <Text>
        Banner title
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
